cppcache/include/geode/DataSerializable.hpp

```diff
@@ -48,14 +51,13 @@ class APACHE_GEODE_EXPORT DataSerializable : public virtual Serializable {
   virtual void fromData(DataInput& dataInput) = 0;
 
   /**
-   * @brief Return the classId of the instance being serialized.
+   * @brief Return the type name of the instance being serialized.
    * This is used by deserialization to determine what instance
    * type to create and deserialize into.
    *
-   * The classId must be unique within an application suite.
-   * Using a negative value may result in undefined behavior.
+   * The type must be unique within an application suite.
    */
-  virtual int32_t getClassId() const = 0;
+  virtual const std::string getType() const { return typeid(*this).name(); }
```

Comment:
Why is this virtual? Does it need to be? Do you plan on allowing derived classes to override it?

Reply:
Also, why not return the `type_info` and use that rather than strings?

```
const std::type_info& getType() const {
    return typeid(*this);
  }
```

Reply:
Shouldn't be virtual.

Returning type_info& doesn't work for managed user classes as it resolves to the exact same value even for different custom classes.

Reply:
Managed types should not influence unmanned types. Managed should be using `Object.GetType()` during mapping and serialization.

Reply:
It became necessary for this to be`virtual` due to the need to make this call in managed and unmanaged code....  both managed/un-managed types end up in the same map

`std::type_info` has a deleted which meant that if we used `Object.getClass()` we would not be able to translate into a `std::type_info` that would be needed inside `cppcache`

`clicache` has the class `ManagedCacheableKeyGeneric` which derives from `native::DataSerializable` and does some marshaling between `clicahe` and `cppcache`
 
We went to using a string return value since ultimately calls to `clicache`'s `TypeRegistry::RegisterType(...)` and `cppcache`'s `SerializationRegistry::AddType(...)` end up in  common map held in `cppcache`

Reply:
I really don't think it should be doing that. Ugh that design was all sorts of broken and shouldn't be dictating design between the two languages. Let me percolate on this.

Reply:
`const std::string` as a return type does not make sense. Each call to this method will allocate a new `std::string` you are then saying it must not be changed so the caller must allocate a new `std::string` to mutate or will accidentally do so if they don't remember to add const or use `auto&&`.

```
# results in single string allocation.
auto&& type = object.getType();

# results in 2 string allocations.
auto type = object.getType();

# results in 2 allocations.
std::string type = object.getType();

std::unordered_map<std::string, int32_t> map;
# results in 2-3 allocations of std::string.
map.emplace(object.getType(), someInt);
```

Reply:
If you eliminate the path where native code needs to access the class information for managed types then these issues can all be resolved with the `const std::type_info&` proposed. When serializing the lookup of type to classId should be happening in the managed layer anyway, so the table for type to classId in the unmanaged code should not be invoked at all. When deserializing we could hook earlier that we are actually managed code and divert to managed code for the type lookups. We do this already for PDX types.